### DISCUSSION TITLE
lazy="warn" loading as a way to log lazy loads, without causing an error

### DISCUSSION THREAD

**Discussion by brouberol** [sqlalchemy/sqlalchemy#9578] [Ideas]
Hi!

I've been toying with an idea recently, and before diving into the SQLAlchemy loader internals, I wanted to probe the community for interest.

## Context

SQLAlchemy provides the [`raiseload` loading technique](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#preventing-unwanted-lazy-loads-using-raiseload) that raises an `InvalidRequestError` exception when a specific relationship tagged with `lazy="raise"` is accessed. This acts as a guarantee that some attributes won't be accessed, and when used with `.raiseload("*")`, this guarantees that only eager-loaded attributes will be accessed, for performance reasons.

However, in the scenario where such an access is made, we usually find out in production, and have to go back and either whitelist an extra relationship, or change the way we access the data in the first place. While that _is_ useful in the long run, the process itself causes disruption, as while some of the invalid accesses could found in unit-tests, others are found via crash reporting tooling.

## Idea 

Could we introduce a `lazy="warn"` lazy loader, that works exactly as `lazy="select"`, except that it logs a warning (vs raising in `lazy="raise"`)? This way, we could tolerate the non-performant data access, but warn about it. It would then be the developer's responsibility to monitor these logs and act upon them to fix the relationship access.

## Example

```python
from sqlalchemy import DeclarativeBase
from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import relationship


BaseModel = DeclarativeBase()


class User(BaseModel):
    __tablename__ = "user"

    id = Column(Integer, primary_key=True)
    name = Column(String(100))
    addresses = relationship(
        Address,
        uselist=True,
        back_populates="user",
        cascade="all, delete-orphan"
    )


class Address(BaseModel):
    __tablename__ = "address"

    id = Column(Integer, primary_key=True)
    email_address = Column(String, nullable=False)
    user_id = Column(Integer, ForeignKey("user_account.id"), nullable=False)
    user = relationship("User", back_populates="addresses")

    def __repr__(self):
        return f"<Address: {self.email_address}>"

```

We could then build queries such as this one, and have the lazy relationship load be logged:

```python
>>> from sqlalchemy.orm import Load
>>> user = session.query(User).options(Load(User).warnload("*")).filter(id=1234).one()
>>> user.address
2023-03-31T18:09:30 [WARNING] - sqlalchemy.orm.strategy options - User.address relationship was lazy loaded 
<Address: balthazar@test.com>
```
 
I'd be happy to hear your thoughts. Thanks!

**Comment by zzzeek** [comment]
so the original idea with lazy loading is that SQLAlchemy is very transparent about SQL logging.  if you observed non-performant parts of your application you'd go through the [profiling steps](https://docs.sqlalchemy.org/en/20/faq/performance.html#how-can-i-profile-a-sqlalchemy-powered-application) to see what queries are being emitted and where, since "lazy loading of relationships" is just one variety of SQL that can lead to slowness - unexpiration of columns, loading of joined inheritance sub-tables, deferred attributes are also examples of SELECTs that are subject to the N+1 problem.

There's a more general goal of improving the logging so that these various queries can be tagged with information as to their origin.   This is currently proposed as #7732 which is a user-only version, but other ORM-initiated queries could also participate in this.

to get the effect of warning for all relationship-level lazy loads, you can implement that with the [do_orm_execute](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) event hook where you'd check [is_relationship_load](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.ORMExecuteState.is_relationship_load) and then to ensure it's a per-object relationship load take a look at [lazy_loaded_from](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.ORMExecuteState.lazy_loaded_from).

Overall I think the goal of warning for different kinds of queries is a good one but I think a more general solution would be more useful since relationship lazy loads are only one of many kinds of query that can be reduced in frequency.

that said, we do have "raise" options for both relationships and columns so it would not be unprecedented for a warning-only version to be implemented as an alternative.

**Reply by phillipuniverse** [reply]
@zzzeek wow, these were some FANTASTIC pointers!! I was able to use this to build a globally-applied disablement of all lazy loading!

```python
from sqlalchemy import event
from sqlalchemy.orm import Session, ORMExecuteState


@event.listens_for(Session, "do_orm_execute")
def receive_do_orm_execute(orm_execute_state: ORMExecuteState):
    if orm_execute_state.is_relationship_load and orm_execute_state.lazy_loaded_from:
        relationship_path = orm_execute_state.loader_strategy_path

        raise ValueError(f"Lazy load detected from {relationship_path}!")
```

This code is designed to raise an exception any time there is a lazy load anywhere in the application. This is helping us globally figure out where we are accidentally loading relationships lazily, without having to depend on individual developers modifying the `relationship()` property appropriately. This helps prevent further footguns!

With your pointers and a little help from the debugger I am soon saying goodbye to my N+1 queries!!

**Reply by zzzeek** [reply]
if you just want to raise, you could as an alternative make sure all of your `relationship()` objects have `lazy="raise"` set on them, that would have the same effect.  it's the "warn" part we dont have built in right now.

**Comment by brouberol** [comment]
Thanks for your response! I'll dig into the pointers you linked, to see whether I can make it work this way.

**Comment by brouberol** [comment]
I managed to implement a simple `Load.warnload` API in just a few lines! 

I didn't end up using signals, as you hinted at, because I wanted that behavior to be _opt-in_ (aka a relationship lazy loading strategy), and not global to the whole app. 

For the record, that looks like this

```python
import sqlalchemy.utils as sqla_utils
from sqlalchemy.orm.strategy_options import loader_option
from sqlalchemy.orm.strategies import LazyLoader
from sqlalchemy.orm.relationships import RelationshipProperty


@RelationshipProperty.strategy_for(lazy="warn")
class CustomLazyLoader(LazyLoader):
    def _emit_lazyload(self, *args, **kwargs):
        property_name = self.parent_property.key
        model_name = self.parent.class_.__name__
        sqla_utils.warn(f"{model_name}.{property_name} was lazy loaded")
        return super(CustomLazyLoader, self)._emit_lazyload(*args, **kwargs)


@loader_option()
def warnload(loadopt, attr):
    return loadopt.set_relationship_strategy(attr, {"lazy": "warn"})
```

This seems to work as intended and explained in the first post.
